File: day11/src/main.rs
use std::fmt::{Display, Formatter};
use regex::Regex;
use crate::Operation::{Multiply, Plus};

enum Operation {
    Multiply(i64),
    Plus(i64),
    Square,
}

impl Clone for Operation {
    fn clone(&self) -> Self {
        match self {
            Multiply(v) => Multiply(*v),
            Operation::Plus(v) => Plus(*v),
            Operation::Square => Operation::Square
        }
    }
}

impl Display for Operation {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Multiply(v) => write!(f, "* {}", v),
            Plus(v) => write!(f, "+ {}", v),
            Operation::Square => write!(f, "* old"),
        }
    }
}

struct Monkey {
    num: usize,
    items: Vec<i64>,
    op: Operation,
    test: i64,
    when_true: i64,
    when_false: i64,
    inspections: usize,
}

impl Monkey {
    pub fn add_item(&mut self, item: i64) {
        self.items.push(item);
    }

    pub fn clear(&mut self) {
        self.items.clear();
    }
}

impl Clone for Monkey {
    fn clone(&self) -> Self {
        return Monkey {
            num: self.num,
            items: self.items.clone(),
            op: self.op.clone(),
            test: self.test,
            when_true: self.when_true,
            when_false: self.when_false,
            inspections: self.inspections,
        };
    }
}


fn parse(input: &Vec<String>) -> Vec<Monkey> {
    let mut result: Vec<Monkey> = vec![];
    let r_monkey = Regex::new(r"^Monkey (\d+):").unwrap();
    let r_items = Regex::new(r"^\s+Starting items: (.*)").unwrap();
    let r_operation = Regex::new(r"^\s+Operation: new = old (.) (\d+|old)").unwrap();
    let r_test = Regex::new(r"^\s+Test: divisible by (\d+)").unwrap();
    let r_true = Regex::new(r"^\s+If true: throw to monkey (\d+)").unwrap();
    let r_false = Regex::new(r"^\s+If false: throw to monkey (\d+)").unwrap();
    let mut i = 0;

    loop {
        if i >= input.len() {
            break;
        }
        let mut l = input.get(i).unwrap();

        let monkey_num = r_monkey.captures(l).unwrap().get(1).unwrap().as_str().parse::<usize>().unwrap();

        i += 1;
        l = input.get(i).unwrap();
        let items = r_items.captures(l).unwrap()
            .get(1).unwrap().as_str()
            .split(", ").collect::<Vec<&str>>()
            .into_iter().map(|s| s.parse().unwrap()).collect();

        i += 1;
        l = input.get(i).unwrap();
        let opc = r_operation.captures(l).unwrap();
        let op = match opc.get(1).unwrap().as_str() {
            "*" => match opc.get(2).unwrap().as_str() {
                "old" => Operation::Square,

                _ => Operation::Multiply(opc.get(2).unwrap().as_str().parse::<i64>().unwrap() as i64)
            },

            "+" => Operation::Plus(opc.get(2).unwrap().as_str().parse::<i64>().unwrap() as i64),

            _ => panic!()
        };

        i += 1;
        l = input.get(i).unwrap();
        let test = r_test.captures(l).unwrap().get(1).unwrap().as_str().parse().unwrap();

        i += 1;
        l = input.get(i).unwrap();
        let when_true = r_true.captures(l).unwrap().get(1).unwrap().as_str().parse().unwrap();

        i += 1;
        l = input.get(i).unwrap();
        let when_false = r_false.captures(l).unwrap().get(1).unwrap().as_str().parse().unwrap();

        result.push(Monkey {
            num: monkey_num,
            items,
            op,
            test,
            when_true,
            when_false,
            inspections: 0,
        });

        i += 2;
    }

    result
}


fn perform_operation(op: &Operation, value: i64) -> i64 {
    return match op {
        Multiply(v) => value * v,
        Operation::Plus(v) => value + v,
        Operation::Square => value * value
    };
}

fn sling_stuff(m: &Monkey) -> Vec<(usize, i64)> {
    m.items.iter().map(|item| {
       let mut new_item = perform_operation(&m.op, *item);

        // println!("Monkey {}", m.num);
        // println!("\t{} -> {}", *item, new_item);

        new_item = new_item / 3;
        // print!("\t Bored to {} ", new_item);
        // print!("\t Divisible by {} ", m.test);

        let target_monkey = if new_item % m.test == 0 {
            // println!("true: Move to {}", m.when_true);
            m.when_true
        } else {
            // println!("false: Move to {}", m.when_false);
            m.when_false
        };

        (target_monkey as usize, new_item)
    }).collect()
}

fn stuff_slinging_simian_shenanigans(monkeys: Vec<Monkey>) -> Vec<Monkey> {
    let mut result: Vec<Monkey> = monkeys.iter().cloned().collect();
    let mut collected_moves: Vec<(usize, i64)> = vec![];

    for m in &mut result {
        for (t, i) in &collected_moves {
            if *t == m.num {
                m.add_item(*i);
            }
        }

        m.inspections += m.items.len();
        collected_moves = collected_moves.into_iter()
            .filter(|(t, _)| m.num != *t).collect();

        let mut moves = sling_stuff(m);
        collected_moves.append(&mut moves);

        m.clear();
    }

    for m in &mut result {
        for (t, i) in &collected_moves {
            if *t == m.num {
                m.add_item(*i);
            }
        }
    }

    result
}


fn main() {
    let data = commons::read_input("input.txt");

    let mut monkeys = parse(&data);

    for i in 0..20 {
        monkeys = stuff_slinging_simian_shenanigans(monkeys);
    }

    for m in &monkeys {
        print!("Monkey {}", m.num);
        println!("\tInspections: {}", m.inspections);

        // println!("\tItems: {}", m.items.iter().map(|x| format!("{x}")).collect::<Vec<String>>().join(", "));
        // println!("\tOperation: new = old {}", match m.op {
        //     Operation::Multiply(o) => format!("* {o}"),
        //     Operation::Plus(o) => format!("+ {o}"),
        //     Operation::Square => String::from("* old")
        // });
        //
        // println!("\tTest: divisible by {}", m.test);
        // println!("\t\tIf true: throw to monkey {}", m.when_true);
        // println!("\t\tIf false: throw to monkey {}", m.when_false);
    }
}

File: day11/js/main.js
const fs = require("fs");

function parse(input) {
    const rmonkey = /^\s*Monkey (\d+):/;
    const ritems = /^\s*Starting items: (.*)/;
    const roperation = /^\s*Operation: new = old (.) (.*)/;
    const rtest = /^\s*Test: divisible by (\d+)/;
    const rif_true = /^\s*If true: throw to monkey (\d+)/;
    const rif_false = /^\s*If false: throw to monkey (\d+)/;

    let res = [];
    let p = 0;

    while (p < input.length) {
        res.push({
            num: 1 * input[p++].match(rmonkey)[1],
            items: input[p++].match(ritems)[1].split(", ").map(i => Number(i)),
            operation: (() => {
                let m = input[p++].match(roperation);
                return {
                    op: m[1],
                    v: m[2]
                }
            })(),
            test: 1 * input[p++].match(rtest)[1],
            ifTrue: 1 * input[p++].match(rif_true)[1],
            ifFalse: 1 * input[p++].match(rif_false)[1],
            inspections: 0
        });
        p++;
    }

    return res;
}

function performOp(op, value) {
    // console.log(value, op);
    switch (op.op) {
        case '+': return value + 1 * op.v;
        case '-': return value - 1 * op.v;
        case '/': return value / 1 * op.v;
        case '*': return op.v === "old" ? value * value : value * 1 * op.v;

        default:
            break;

    }
}

function stuffSlingingSimianShenanigans(monkeys) {

    for(let i = 0; i < monkeys.length; i++) {

        const m = monkeys[i];
        console.log(m.num, m.items.length, "->", `[${m.items.join(", ")}]`);

        m.inspections += m.items.length;

        if (m.num === 6) {
            console.log(m.items.join(", "))
        }

        m.items = m.items.map(i => performOp(m.operation, i));

        if (m.num === 6) {
            console.log(m.items.join(", "))
        }
        m.items.forEach(i => {
            let ni = Math.floor(i / 3);

            if (ni === 26019075) {
                console.log(m.operation)
                console.log(i);
                console.log("---------->", m.num, m.test);
            }

            if (ni % m.test === 0) {
                monkeys[m.ifTrue].items.push(ni);
            } else {
                monkeys[m.ifFalse].items.push(ni);
            }
        });

        m.items = [];
    }

    return monkeys;
}


let input = fs.readFileSync("../input.txt");

let monkeys = parse(("" + input).split("\r\n"));

for (i = 0; i < 20; i++) {
    console.log("\n\nRound", i + 1);
    monkeys = stuffSlingingSimianShenanigans(monkeys);
}


console.log(monkeys.map(m => {
    return { num: m.num, inspections: m.inspections}
}))

monkeys.sort((a, b) => b.inspections - a.inspections);

console.log(monkeys[0].inspections * monkeys[1].inspections);